File: java/other/SnowFlake.java
package other;

public class SnowFlake {
    private static final long _startTimeStamp = 1356019200000L;

    private long _timeStamp = 0;
    private int _dataCenterId = 0;
    private int _workerId = 0;
    private int _sequence = 0;

    public SnowFlake(int dataCenterId, int workerId) {
        _dataCenterId = dataCenterId;
        _workerId = workerId;
    }

    public long uuid() {
        long tick = System.currentTimeMillis();
        if (tick < _timeStamp) {
            return Integer.MIN_VALUE;
        }

        if (tick == _timeStamp) {
            _sequence = (_sequence + 1) & (~(-1 << 12));
            if (0 == _sequence) {
                while (tick == _timeStamp) {
                    tick = System.currentTimeMillis();
                }
            }
        } else {
            _sequence = 0;
        }
        _timeStamp = tick;
        return ((tick - _startTimeStamp) << 22) | (_dataCenterId << 17) | (_workerId << 12) | _sequence;
    }
}

File: py/other/SnowFlake.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import sys, time

class SnowFlake:
    def __init__(self, dataCenterId: int, workerId: int):
        self.__startTimeStamp = 1356019200000
        self.__timeStamp, self.__dataCenterId, self.__workerId, self.__sequence = 0, dataCenterId, workerId, 0

    def uuid(self) -> int:
        tick = int(time.time() * 1000)
        if tick < self.__timeStamp:
            return -sys.maxsize

        if tick == self.__timeStamp:
            self.__sequence = (self.__sequence + 1) & (~(-1 << 12))
            if not self.__sequence:
                while tick == self.__timeStamp:
                    tick = int(time.time() * 1000)
        else:
            self.__sequence = 0
        self.__timeStamp = tick
        return ((tick - self.__startTimeStamp) << 22) | (self.__dataCenterId << 17) | (self.__workerId << 12) | self.__sequence
